project/backend/src/test/java/com/google/sps/JsonToRestaurantParserTest.java

```diff
@@ -0,0 +1,112 @@
+package com.google.sps;
+
+import static com.google.common.truth.Truth.assertThat;
+import com.google.sps.data.JsonToRestaurantParser;
+import com.google.sps.data.Restaurant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class JsonToRestaurantParserTest {
+
+  private static final int MISSING_FIELD = -1;
+  private static final int PRICE_LEVEL = 2;
+  private static final int NUM_RATINGS = 415;
+  private static final double RATING = 3.78;
+  private static final List<String> EMPTY_TYPES = new ArrayList<>();
+  private static final List<String> ALL_TYPES = Arrays.asList(new String[]{"meal_takeaway", "meal_delivery", "restaurant"});
+  private static final String RESTAURANT_NAME = "RESTAURANT_A";
+  private static final String PLACE_ID = "A12345";
+  private static final String VICINITY = "10 Main Street";
+  private static final Map<String, Double> COORDS = new HashMap<String, Double>() {{
+    put("lat", 10.0);
+    put("lng", 20.0);
+  }};
+
+  @Test
+  public void allRequiredAndOptionalFields() throws JSONException {
+    // JSON object has all required and all optional fields.
+    // We should see a restaurant with all the values in the JSON object.
+    JSONObject geometry = new JSONObject().put("location", COORDS);
+    JSONObject body = new JSONObject()
+      .put("place_id", PLACE_ID)
+      .put("name", RESTAURANT_NAME)
+      .put("vicinity", VICINITY)
+      .put("types", EMPTY_TYPES)
+      .put("geometry", geometry)
+      .put("rating", RATING)
+      .put("user_ratings_total", NUM_RATINGS)
+      .put("price_level", PRICE_LEVEL);
+
+    Restaurant expected = Restaurant.create(
+      PLACE_ID, RESTAURANT_NAME, VICINITY, COORDS, RATING, NUM_RATINGS,
+      PRICE_LEVEL, EMPTY_TYPES);
+    Restaurant actual = JsonToRestaurantParser.toRestaurant(body);
+    assertThat(actual).isEqualTo(expected);
+  }
+
+  @Test
+  public void extraneousTypes() throws JSONException {
+    // JSON object has types that should be disregarded.
+    // We should see a restaurant with only the three relevant types.
+    List<String> randomTypes = Arrays.asList(new String[]{"type a", "meal_takeaway", "type b", "meal_delivery", "restaurant"});
+    JSONObject geometry = new JSONObject().put("location", COORDS);
+    JSONObject body = new JSONObject()
+      .put("place_id", PLACE_ID)
+      .put("name", RESTAURANT_NAME)
+      .put("vicinity", VICINITY)
+      .put("types", randomTypes)
+      .put("geometry", geometry);
+
+    Restaurant actual = JsonToRestaurantParser.toRestaurant(body);
+    assertThat(actual.getPlaceTypes()).containsExactlyElementsIn(ALL_TYPES);
+  }
+
+  @Test
+  public void noOptionalFields() throws JSONException {
```

Comment:
so for these cases consider a naming pattern along the lines of 
situationBeingTested_expectedOutcome 
-- in this specific case, that could translate to 
noOptionalFields_usesDefaultsForOptional 
^ which eliminates the need for "what this test is doing" comments and makes it self-explanatory 

This feels like it can get verbose with test naming and setup and that sort of thing is more ok in unit testing than actual code. On a related note - we often talk about DRY (dont repeat yourself) when we refer to actual code where we encourage helper methods and abstractions when needed to avoid repitition. 
But for tests, often, DRY is not encouraged https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests. -- and instead we choose Descriptive and meaningful phrases 

of course, there's no one size fits all so usually tests end up being on the spectrum of dry --> damp but in general with unit testing, some repetition (such as with setting up inputs in a given test etc) is not just fine but encouraged as we usually want to make tests as self-contained as possible with distinct "setup" "do the work" "assert the outcome" blocks

Reply:
So if I use that naming pattern, should I do it for all the tests or only the ones that it seems necessary for? For example the noFields test seems self explanatory of what is expected to happen so would I want to still follow that naming convention there?

Reply:
yep - only ones where the expected outcome isn't obvious without comments - noFields seems fine as is!